### PR TITLE
feat: Goblin Honk Composer/Prover/Verifier

### LINKED ISSUE #616: Ultra Goblin Composer

### LINKED ISSUE #617: Ultra Goblin Prover/Verifier

### PR DESCRIPTION
**Pull request by ledwards2225**
# Description

The Composer, Prover and Verifier for Goblin Ultra Honk. These are implemented as instantiations of the existing "Ultra" template classes, e.g. `GoblinUltraComposer = UltraComposer_<flavor::GoblinUltra>`, etc.

This work makes it possible to create and verify a Goblin Ultra Honk proof. "Verification" here means confirmation that the ECC op gates have been correctly incorporated into the circuit (checked via a new relation) plus genuine verification of the conventional UltraHonk proof. The _correctness_ of the the ECC op gates is of course not verified, as that is the role of the other components of the Goblin stack.

The new "op gate consistency" relation simply checks that the op wire values have been correctly copied into the conventional wires (needed for copy constraints) and that the op wire polynomials vanish everywhere outside the range of ECC op gates.

# Checklist:

- [ ] I have reviewed my diff in github, line by line.
- [ ] Every change is related to the PR description.
- [ ] The branch has been merged with/rebased against the head of its merge target.
- [ ] There are no unexpected formatting changes, superfluous debug logs, or commented-out code.
- [ ] There are no circuit changes, OR a cryptographer has been assigned for review.
- [ ] New functions, classes, etc. have been documented according to the doxygen comment format. Classes and structs must have `@brief` describing the intended functionality.
- [ ] If existing code has been modified, such documentation has been added or updated.
- [ ] No superfluous `include` directives have been added.
- [ ] I have [linked](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue) to any issue(s) it resolves.
- [ ] I'm happy for the PR to be merged at the reviewer's next convenience.

### CODE REVIEW

File: cpp/src/barretenberg/honk/sumcheck/relations/relation_correctness.test.cpp [hunk 1]
```diff
@@ -60,6 +61,145 @@ template <typename Flavor> void check_relation(auto relation, auto circuit_size,
     }
 }
 
+template <typename Flavor> void create_some_add_gates(auto& circuit_builder)
```
Comment (ledwards2225) [inline cpp/src/barretenberg/honk/sumcheck/relations/relation_correctness.test.cpp:64]: For some reason the diff for this test file is hard to read. All I've done is (1) modularize the creation of various types of gates (arithmetic, lookup, etc.) into individual functions for reuse, and (2) add a relation correctness test for the GoblinUltraComposer which includes ECC op gates and checks the new op wire consistency relation in addition to the nominal Ultra Honk relations.

File: cpp/src/barretenberg/honk/composer/ultra_composer.cpp [hunk 2]
```diff
@@ -137,10 +130,11 @@ template <UltraFlavor Flavor> void UltraComposer_<Flavor>::compute_witness(Circu
     // Copy memory read/write record data into proving key. Prover needs to know which gates contain a read/write
     // 'record' witness on the 4th wire. This wire value can only be fully computed once the first 3 wire polynomials
     // have been committed to. The 4th wire on these gates will be a random linear combination of the first 3 wires,
-    // using the plookup challenge `eta`. We need to update the records with an offset Because we shift the gates by
-    // the number of public inputs plus an additional offset for a zero row.
+    // using the plookup challenge `eta`. We need to update the records with an offset Because we shift the gates to
+    // account for everything that comes before them in the execution trace, e.g. public inputs, a zero row, etc.
     auto add_public_inputs_offset = [this](uint32_t gate_index) {
-        return gate_index + num_public_inputs + zero_row_offset;
+        size_t offset = num_ecc_op_gates + num_public_inputs + num_zero_rows;
```
Comment (Rumata888) [inline cpp/src/barretenberg/honk/composer/ultra_composer.cpp]: You are recalculating this value every time you call this function. If you take it out, you won't need *this* and you will just add the offset to the indices without recomputing the offset

File: cpp/src/barretenberg/proof_system/composer/permutation_lib.hpp [hunk 3]
```diff
@@ -96,6 +94,32 @@ std::vector<CyclicPermutation> compute_wire_copy_cycles(const typename Flavor::C
...
+        // Set ecc op gate offset and update offsets for PI and conventional gates
+        const size_t op_gates_offset = num_zero_rows;
+        const size_t num_ecc_op_gates = circuit_constructor.num_ecc_op_gates;
+        pub_inputs_offset += num_ecc_op_gates;
+        gates_offset += num_ecc_op_gates;
+
+        const auto& op_wires = circuit_constructor.ecc_op_wires;
+        // Iterate over all variables of the ecc op gates, and add a corresponding node to the cycle for that variable
+        for (size_t i = 0; i < num_ecc_op_gates; ++i) {
+            // Note: We exclude the first op wire since it contains the op codes which are not stored in variables
```
Comment (Rumata888) [inline cpp/src/barretenberg/proof_system/composer/permutation_lib.hpp:113]: Are you sure about that? You'll have to fix the operation in some way or other. One of the ways to do that is to create a cycle, connecting all the same operations together and to a known constant. Without that mechanism you'll have to fix them with some other relation.

**Review by Rumata888 — APPROVED** [review]
Everything looks great. The only thing that I think could be an issue in the long run is the OP part not being included in copy constraints, but maybe I missed something there. Anyway, should be an easy fix. LGTM

File: cpp/src/barretenberg/honk/composer/ultra_composer.cpp (hunk 2 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/honk/composer/ultra_composer.cpp]: Good catch, thanks. Moved it out

File: cpp/src/barretenberg/proof_system/composer/permutation_lib.hpp (hunk 3 above)
Comment (ledwards2225) [reply cpp/src/barretenberg/proof_system/composer/permutation_lib.hpp:113]: Good point. I'm not sure what the best approach is. I like the idea of using a copy constraint to constrain to constant values. As you say, it should be easy enough to retrofit a solution. I added a TODO that reflects your suggestions.

**Comment by ledwards2225** [comment]
Completed as PR into aztec-packages (#1220)

**Pull request closed without merging**